### PR TITLE
Fix Assertion error in BatchNorm2d 

### PR DESCRIPTION
**Pull request by kamalrajkannan78**
## Description

- This PR fixes [#500](https://github.com/tenstorrent/tt-forge-fe/issues/500)

## Logs

- [BM_cases_before_fix.log](https://github.com/user-attachments/files/17485943/BM_cases_before_fix.log)
- [BM_cases_after_fix.log](https://github.com/user-attachments/files/17485942/BM_cases_after_fix.log)

### CODE REVIEW

File: forge/test/mlir/test_ops.py [hunk 1]
```diff
@@ -14,6 +14,30 @@
...
+@pytest.mark.parametrize(
+    "num_channels, height, width",
+    [
+        (32, 56, 56),
+        (64, 112, 112),
+        (128, 224, 224),
+        (256, 28, 28),
+    ],
+)
+def test_multidim_unsqueeze(num_channels, height, width):
```
Comment (nvukobratTT) [inline forge/test/mlir/test_ops.py:26]: Can we add a specific unit test that will target expend_dims only? Feel free to keep batch norm test as well, but let's add additional one to target only the expend_dim op :))

File: forge/test/mlir/test_ops.py [hunk 2]
```diff
@@ -14,6 +14,30 @@
...
+
+    framework_model = nn.BatchNorm2d(num_features=num_channels)
+
+    inputs = [torch.randn(1, num_channels, height, width)]
```
Comment (nvukobratTT) [inline forge/test/mlir/test_ops.py:30]: For batch norm, can we also test multiple batches?

File: forge/test/mlir/test_ops.py (hunk 1 above)
Comment (kamalrajkannan78) [reply forge/test/mlir/test_ops.py:26]: - I didn't find any torch operation that gives our required relay expression (expand_dims with num_newaxis=2)
` %4 = expand_dims(%3, axis=1, num_newaxis=2) /* ty=Tensor[(32, 1, 1), float32] */`

- Only way to get expand_dims with num_newaxis=2, [view operation should be applied on const expression](https://github.com/tenstorrent/tt-tvm/blob/2156fc1401cc8491712f25882981bac762800dc5/python/tvm/relay/frontend/pytorch.py#L1834) in nn.module like below 

```
def test_check():
    class Multidim_unsqueeze(nn.Module):
        def __init__(self):
            super().__init__()
            self.ip2 = torch.rand(128,128)

        def forward(self, tensor):
            return tensor + self.ip2.view(1,1,128,128)
    
    inputs = [torch.randn(128,128)]
    framework_model = Multidim_unsqueeze()
    fw_out = framework_model(*inputs)

    compiled_model = forge.compile(framework_model, sample_inputs=inputs)
    co_out = compiled_model(*inputs)
```

- but the problem here is view operation applied on const expression, so it is removed from relay expression after After DynamicToStatic relay pass.

**Before DynamicToStatic relay pass.**

```
{I.GlobalVar("main"): fn (%tensor: Tensor[(128, 128), float32] /* ty=Tensor[(128, 128), float32] span=aten::add_0.tensor:0:0 */) -> Tensor[(1, 1, 128, 128), float32] {
  %0 = expand_dims(meta[relay.Constant][0] /* ty=Tensor[(128, 128), float32] span=aten::view_0:0:0 */, axis=-1, num_newaxis=2) /* ty=Tensor[(128, 128, 1, 1), float32] span=C.graph: aten::view, jit._trace.TopLevelTracedModule: test_ops.Multidim_unsqueeze:::0:0 */;
  %1 = reshape(%0, newshape=[1, 1, 128, 128]) /* ty=Tensor[(1, 1, 128, 128), float32] span=C.graph: aten::view, jit._trace.TopLevelTracedModule: test_ops.Multidim_unsqueeze:::0:0 */;
  add(%tensor, %1) /* ty=Tensor[(1, 1, 128, 128), float32] span=C.graph: aten::add, jit._trace.TopLevelTracedModule: test_ops.Multidim_unsqueeze:::0:0 */
} /* ty=fn (Tensor[(128, 128), float32]) -> Tensor[(1, 1, 128, 128), float32] */
}
```
**After DynamicToStatic relay pass.**
```
{I.GlobalVar("main"): fn (%tensor: Tensor[(128, 128), float32] /* ty=Tensor[(128, 128), float32] span=aten::add_0.tensor:0:0 */) -> Tensor[(1, 1, 128, 128), float32] {
  add(%tensor, meta[relay.Constant][0] /* ty=Tensor[(1, 1, 128, 128), float32] */) /* ty=Tensor[(1, 1, 128, 128), float32] span=C.graph: aten::add, jit._trace.TopLevelTracedModule: test_ops.Multidim_unsqueeze:::0:0 */
} /* ty=fn (Tensor[(128, 128), float32]) -> Tensor[(1, 1, 128, 128), float32] */
}
```

- Please refer log for better clarity - [expand_dims_issue.log](https://github.com/user-attachments/files/17490358/oct23_zzz_check_2.log)

File: forge/test/mlir/test_ops.py (hunk 2 above)
Comment (kamalrajkannan78) [reply forge/test/mlir/test_ops.py:30]: - Observed pcc drop when Batch size >1.

| Input Shape         | PCC              |
|---------------------|------------------------|
| (2, 64, 112, 112)   | 0.6199620538910243      |
| (4, 64, 28, 28)     | 0.4935656199688308      |
| (8, 64, 112, 112)   | 0.40443518583193394     |
| (16, 128, 224, 224) | -0.0004391043640747615  |
| (32, 256, 28, 28)   | 0.39200606381500713     |

Comment (nvukobratTT) [reply forge/test/mlir/test_ops.py:30]: Make sense, let's focus on batch 1, but please leave a comment in the test these details regarding PCC.

File: forge/test/mlir/test_ops.py (hunk 1 above)
Comment (nvukobratTT) [reply forge/test/mlir/test_ops.py:26]: I see, can you test out can we mimic it with simple reshape?

Comment (kamalrajkannan78) [reply forge/test/mlir/test_ops.py:26]: - [Reshape is mapped to reshape itself](https://github.com/tenstorrent/tt-forge-fe/blob/9cae62d8869590355342179c04f883b74aedb1ef/forge/forge/tvm_to_python.py#L434).

```
class Multidim_unsqueeze(nn.Module):
    def __init__(self):
        super().__init__()

    def forward(self, tensor):
        return torch.reshape(tensor,(1,1,128,128))


inputs = [torch.randn(128,128)]
framework_model = Multidim_unsqueeze()
fw_out = framework_model(*inputs)
```

 After front end conversion , above code yeilds

```
{I.GlobalVar("main"): fn (%tensor: Tensor[(128, 128), float32] /* ty=Tensor[(128, 128), float32] span=aten::reshape_0.tensor:0:0 */) -> Tensor[(1, 1, 128, 128), float32] {
  reshape(%tensor, newshape=[1, 1, 128, 128]) /* ty=Tensor[(1, 1, 128, 128), float32] span=C.graph: aten::reshape, jit._trace.TopLevelTracedModule: test_ops.Multidim_unsqueeze:::0:0 */
} /* ty=fn (Tensor[(128, 128), float32]) -> Tensor[(1, 1, 128, 128), float32] */}
```

After relay passes ,

```
def @main(%tensor: Tensor[(128, 128), float32] /* ty=Tensor[(128, 128), float32] span=aten::reshape_0.tensor:0:0 */) -> Tensor[(1, 1, 128, 128), float32] {
  reshape(%tensor, newshape=[1, 1, 128, 128]) /* ty=Tensor[(1, 1, 128, 128), float32] span=C.graph: aten::reshape, jit._trace.TopLevelTracedModule: test_ops.Multidim_unsqueeze:::0:0 */
}
```

As our relay expression having name **reshape** not **expand dims** it will not pass through [our logic](https://github.com/tenstorrent/tt-tvm/pull/39/files), so this unit test will not test our logic. 

- [Reshape_to_expand_dims.log](https://github.com/user-attachments/files/17495543/oct23_zzz_3.log)

Comment (nvukobratTT) [reply forge/test/mlir/test_ops.py:26]: I see, thanks for checking it out!

**Review by nvukobratTT — APPROVED** [review]
